### ISSUE TITLE
Improper Handling of Search Params within Navigators

### ISSUE THREAD

**Issue by Braden1996** [expo/router#232]
### Summary

Hello 👋

I've recently tried porting our production app to expo-router, as unifying the navigation mental model with that of the web is appealing to us.

However, I've run into quite a few seemingly blocking issues:

1. [Ancestor search params are not available](https://github.com/expo/router/issues/224)
2. Navigating away from a dynamic route within a stack crashes the app, as the screen we're departing from can no longer access its dynamic search param.
   - During navigation transition, React Navigation continues to render the departing screen with its initial parameters.
   - Expo Router seems to try to render the departing screen with the new parameters, which causes the entire thing to crash.

Whilst this issue is for the 2nd case, I feel there is as wider incompatibility with how React Navigation thinks about search parameters vs how Expo Router does. Are there any thoughts around this, or a fix in the pipeline? I'm debating rolling back the upgrade for now, as this is a blocking issue and seems quite fundamental.

Cheers :)

### Minimal reproducible example

I have created a basic repository of this issue here: https://github.com/Braden1996/expo-router-bugs
Here is a short video demonstrating the issue: 

https://user-images.githubusercontent.com/5165963/215512039-47de4dc6-9945-4b77-8ec0-1c8fc448ca67.mp4

**Comment by Braden1996** [comment]
For the moment, I am circumventing this particular issue by maintaining a cache of prior search param values. The cached value can only be updated by the route which instantiated it. This isn't the desired behaviour, but it's enough for my desired use-case.

```
export function useSearchParam<V extends string>(key: RouterSegments): V | undefined {
    const params = useSearchParams()
    const param = params[key] as V | undefined

    const lastParam = React.useRef<{pathnamePrefix: string; param: V} | undefined>(undefined)
    const segments = useSegments()
    const dynamicSegment = `[${key}]`
    if (segments.includes(dynamicSegment)) {
        const pathnamePrefix = segments.join("/")
        if (
            param !== undefined &&
            (lastParam.current === undefined || lastParam.current.pathnamePrefix === pathnamePrefix)
        ) {
            lastParam.current = {pathnamePrefix, param}
        }
    }

    return lastParam.current?.param
}

export function useAssertSearchParam<V extends string>(key: RouterSegments) {
    const param = useSearchParam<V>(key)

    if (param === undefined) {
        throw new Error(`Missing value for navigation param: '${key}'.`)
    }

    return param
}
```

**Comment by EvanBacon** [comment]
This seems like a duplicate of your other issue https://github.com/expo/router/issues/224

**Comment by Braden1996** [comment]
> This seems like a duplicate of your other issue #224

It's definitely pretty similar, as they both relate to the use of navigators and dynamic routes; although the actual encountered problem is a bit different.

In my other issue (https://github.com/expo/router/issues/224), the problem is that `useSearchParams` doesn't always contain the ancestor segment values. Practically this meant we'd hit error when a tab navigator was nested within a dynamic segment (as not all tabbed screen could access the common segment).

In contrast, this issue is relating to Stack navigators. If one of the Stack's children routes contained a dynamic segment, then we'd hit an error when navigating away from that screen as there would be a render pass of the old screen's components but with the new destination pathname.

For example, imagine we had the routing structure:
```
<Stack>
    <Stack.Screen name="index" />
    <Stack.Screen name="list/[listId]" />
    <Stack.Screen name="item/[itemId]" />
</Stack>
```
Navigating from `list/[listId]` to `item/[itemId]` would throw an error, as the `list/[listId]`'s screen performs a render without its segment value in the current pathname.

I've not dug into Expo Router enough to know if these two issues could be resolved by the same internal change. Happy to fold this into my initial one if that works better for you! :)

**Comment by EvanBacon** [comment]
Resolved in 1.0.0-rc5 with `useLocalSearchParams`
